algorithm/cgal/cgal.cc

```diff
@@ -3840,6 +3846,287 @@ int DisjointSurfaceMeshesWithIncrementalMasking(int meshCount,
+  for (int nth = 0; nth < input_count; nth++) {
+    is_masked[nth] = getIsMasked(nth).as<bool>();
+    switch (types[nth]) {
+      case GEOMETRY_MESH: {
+        std::unique_ptr<Surface_mesh> oriented_mesh(
+            new Surface_mesh(*geometry->getMesh(nth)));
+        CGAL::Polygon_mesh_processing::transform(
+            *transforms[nth], *oriented_mesh, CGAL::parameters::all_default());
+        if (IsPlanarSurfaceMesh(planes[nth], *oriented_mesh)) {
+          geometry->setType(nth, GEOMETRY_POLYGONS_WITH_HOLES);
+          types[nth] = GEOMETRY_POLYGONS_WITH_HOLES;
+          for (int nth_plane = 0; nth_plane < nth; nth_plane++) {
+            if (types[nth_plane] == GEOMETRY_POLYGONS_WITH_HOLES &&
+                IsCoplanarSurfaceMesh(planes[nth_plane], *oriented_mesh)) {
+              // Force a consensus on plane choices.
+              planes[nth] = planes[nth_plane];
+              break;
+            }
+          }
+          PlanarSurfaceMeshToPolygonSet(planes[nth], *oriented_mesh, *gps[nth]);
+          bounds_2[nth] = computePolygonSetBounds(*gps[nth]);
+        } else {
+          bounds_3[nth] = CGAL::Polygon_mesh_processing::bbox(*oriented_mesh);
+          geometry->setOutputMesh(nth, oriented_mesh);
+          break;
+        }
+      }
+      case GEOMETRY_POLYGONS_WITH_HOLES: {
+        bounds_2[nth] = computePolygonSetBounds(*gps[nth]);
+        break;
+      }
+      default:
+        break;
+    }
+  }
+
+  for (int start = input_count - 2; start >= 0; start--) {
+    switch (types[start]) {
+      case GEOMETRY_MESH: {
+        if (CGAL::is_empty(*meshes[start])) {
+          continue;
+        }
+        for (int nth = start + 1; nth < input_count; nth++) {
+          if (is_masked[nth] || types[nth] != GEOMETRY_MESH ||
+              CGAL::is_empty(*meshes[nth]) ||
+              !CGAL::do_overlap(bounds_3[start], bounds_3[nth])) {
+            continue;
+          }
+          Surface_mesh cut(*meshes[nth]);
+          if (!CGAL::Polygon_mesh_processing::corefine_and_compute_difference(
+                  *meshes[start], cut, *meshes[start],
+                  CGAL::parameters::all_default(),
+                  CGAL::parameters::all_default(),
+                  CGAL::parameters::all_default())) {
+            return STATUS_ZERO_THICKNESS;
+          }
+          demesh(*meshes[start]);
+          bounds_3[start] = CGAL::Polygon_mesh_processing::bbox(*meshes[start]);
+        }
+        break;
+      }
+      case GEOMETRY_POLYGONS_WITH_HOLES: {
+        for (int nth = start + 1; nth < input_count; nth++) {
+          // TODO: Disjunction of planar by volume.
+          if (is_masked[nth] || types[nth] != GEOMETRY_POLYGONS_WITH_HOLES ||
+              planes[start] != planes[nth] ||
+              !CGAL::do_overlap(bounds_2[start], bounds_2[nth])) {
+            continue;
+          }
+          gps[start]->difference(*gps[nth]);
+        }
+        // TODO: Handle disjunction of surface by volume.
```

Comment:
Will cgal be able to do these? I don't remember seeing anything in there relating to interacting 2d and 3d representations...Although I guess the volume could always be sectioned

Reply:
Well, it's not really 2d vs 3d.

You can clip or cut open meshes with closed meshes.

But, yeah -- otherwise you can take a section, which we can convert to polygons-with-holes, which is probably the right approach here.